### PR TITLE
[Bug] Fix for "Moves Can Miss Against Protect, Baneful Bunker, King's Shield" 

### LINKED ISSUE #4060: [Bug] Moves Can Miss Against Protect, Baneful Bunker, King's Shield
### Describe the bug

Moves may miss against protect and protect clones. 

### Session export file

_No response_

### User data export file

_No response_

### Expected behavior

https://replay.pokemonshowdown.com/gen9customgame-2195472381-exy2nnl4kxvx63gpep5tsp60al6a7lepw
Whether a move is protected against should be considered before accuracy

### Screenshots / Videos

-6 acc Nidoran misses against Cyndaquil's protect.

https://github.com/user-attachments/assets/ee7151da-9106-4f9f-9d4a-241387e5ecd2

### What OS did you observe the bug on?

Other

### If other please specify

gameplay

### Which browser do you use?

Other

### If other please specify

gameplay

### Additional context

Protection moves that apply their effect after a move is protected against would not trigger their effects. The protection move retains the lowered success rate, so in effect, this can drastically lower the reliability of protection moves.

Worth mentioning
https://github.com/pagefaultgames/pokerogue/issues/3673
Where moves can miss against mons with immunity abilities, if when acc is checked for would be modified.

### PR DESCRIPTION
**Pull request by PrabbyDD**
## What are the changes the user will see?
Fixing  #4060, so if pokemon misses against a move that protects, it will properly protect against the move despite missing
After effects of a protected, missed attack do t trigger if the move is contact (tested with slash)
After effects of a protected, missed attack do not trigger if the move is non contact (tested with flash cannon)

## Why am I making these changes?
To fix  #4060 

## What are the changes from a developer perspective?
Moved protect check to be before hit check, and avoid hitcheck if the pokemon is protected

### Screenshots/Videos

https://github.com/user-attachments/assets/bdd9c2b8-1388-45e0-b6f1-0ec8cb605235

https://github.com/user-attachments/assets/8417e7d0-c2df-4f94-943b-f0710e56b743

https://github.com/user-attachments/assets/c75280d4-4297-49fa-b50e-bb3cc3ac0846

https://github.com/user-attachments/assets/2cb2324c-257f-4fb8-bc19-62f868e31b2e

https://github.com/user-attachments/assets/0bc2c81d-3338-49c3-9ee4-feaef88f1976

## How to test the changes?
Use npm run test, as well as changing override files to test moves, and manually go into move-effect-phase.ts and change return value of hitCheck() to be false. 

npm test obstruct (modified)
npm test baneful_bunker (added)

Run these two tests as specifically. I also tested manually via removing the forceMiss() function, and changing hitCheck() to return false to be sure. 

I would appreciate any feedback on the changes to the move-effect-phase.ts file. Not really a good way to test protect specifically, since it doesnt have after effects. 

I tested manually with other moves as well, like King's Shield and Burning Bulwark. Seems to work fine for all now. 

## Checklist
- [X] **I'm using `beta` as my base branch**
- [x ] There is no overlap with another PR?
- [ x] The PR is self-contained and cannot be split into smaller PRs?
- [ x] Have I provided a clear explanation of the changes?
- [ x] Have I considered writing automated tests for the issue?
- [ ] If I have text, did I make it translatable and add a key in the English locale file(s)?
- [ x] Have I tested the changes (manually)?
    - [x ] Are all unit tests still passing? (`npm run test`)
- [ x] Are the changes visual?
  - [ x] Have I provided screenshots/videos of the changes?

### CODE REVIEW

File: src/test/moves/baneful_bunker.test.ts [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+    game.override.enemySpecies(Species.SNORLAX);
+    game.override.enemyAbility(Abilities.INSOMNIA);
+    game.override.enemyMoveset(Moves.BANEFUL_BUNKER);
+
+    game.override.startingLevel(100);
+    game.override.enemyLevel(100);
+  });
+  test(
+    "Baneful Bunker test",
```
Comment (innerthunder) [inline src/test/moves/baneful_bunker.test.ts:41]: Please use more descriptive names for tests:
- "should protect the user and poison attackers that make contact"
- "should apply effects regardless of accuracy checks"
- "should not poison attackers that don't make contact"

File: src/test/moves/baneful_bunker.test.ts [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    async () => {
+      await game.classicMode.startBattle([Species.CHARIZARD]);
+
+      const leadPokemon = game.scene.getPlayerPokemon()!;
+      const enemyPokemon = game.scene.getEnemyPokemon()!;
+
+      game.move.select(Moves.SLASH);
+      await game.setTurnOrder([BattlerIndex.ENEMY, BattlerIndex.PLAYER]);
+      await game.phaseInterceptor.to("MoveEffectPhase");
+      await game.phaseInterceptor.to("TurnEndPhase");
```
Comment (innerthunder) [inline src/test/moves/baneful_bunker.test.ts:51]:
```suggestion
      await game.phaseInterceptor.to("BerryPhase", false);
```
Just awaiting until after all moves in the turn are resolved is sufficient, and this avoids the possibility of the enemy activating a randomly generated Lum Berry.

File: src/test/moves/baneful_bunker.test.ts [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+      const leadPokemon = game.scene.getPlayerPokemon()!;
+      const enemyPokemon = game.scene.getEnemyPokemon()!;
+
+      game.move.select(Moves.SLASH);
+      await game.setTurnOrder([BattlerIndex.ENEMY, BattlerIndex.PLAYER]);
+      await game.phaseInterceptor.to("MoveEffectPhase");
+
+      await game.move.forceMiss();
+      await game.phaseInterceptor.to("TurnEndPhase");
```
Comment (innerthunder) [inline src/test/moves/baneful_bunker.test.ts:69]:
```suggestion
      await game.phaseInterceptor.to("BerryPhase", false);
```

File: src/test/moves/baneful_bunker.test.ts [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+
+      const leadPokemon = game.scene.getPlayerPokemon()!;
+      const enemyPokemon = game.scene.getEnemyPokemon()!;
+
+      game.move.select(Moves.FLASH_CANNON);
+      await game.setTurnOrder([BattlerIndex.ENEMY, BattlerIndex.PLAYER]);
+      await game.phaseInterceptor.to("MoveEffectPhase");
+
+      await game.move.forceMiss();
+      await game.phaseInterceptor.to("TurnEndPhase");
```
Comment (innerthunder) [inline src/test/moves/baneful_bunker.test.ts:89]:
```suggestion
      await game.phaseInterceptor.to("BerryPhase", false);
```

File: src/test/moves/obstruct.test.ts [hunk 5]
```diff
@@ -43,6 +43,23 @@ describe("Moves - Obstruct", () => {
     expect(enemy.getStatStage(Stat.DEF)).toBe(-2);
   }, TIMEOUT);
 
+  it("obstruct miss test, should still reduce defense despite missed attack", async () => {
```
Comment (innerthunder) [inline src/test/moves/obstruct.test.ts:46]:
```suggestion
  it("bypasses accuracy checks when applying protection and defense reduction", async () => {
```
Using the same naming/style conventions as the other tests in this file

**Review by innerthunder — COMMENTED** [review]
The source code seems good; just some nits on the test code

File: src/test/moves/baneful_bunker.test.ts (hunk 1 above)
Comment (PrabbyDD) [reply src/test/moves/baneful_bunker.test.ts:41]: What does nit mean?

Comment (innerthunder) [reply src/test/moves/baneful_bunker.test.ts]: short for nit picks/nit picking -- basically small issues with the code that don't impact behavior (like style choices, comments, things like that)

**Review by innerthunder — APPROVED** [review]

**Pull request merged**